fdb-extensions/src/main/java/com/apple/foundationdb/async/rabitq/RaBitQuantizer.java

```diff
@@ -0,0 +1,361 @@
+     *                 This method internally uses |r| normalized to unit L2.
+     */
+    private QuantizeExResult exBitsCode(@Nonnull final RealVector residual) {
+        int dims = residual.getNumDimensions();
+
+        // oAbs = |r| normalized (RaBitQ does this before quantizeEx)
+        final RealVector oAbs = absOfNormalized(residual);
+
+        final QuantizeExResult q = quantizeEx(oAbs);
+
+        int[] k = q.code;
+        // revert codes for negative dims
+        int[] signed = new int[dims];
+        int mask = (1 << numExBits) - 1;
+        for (int j = 0; j < dims; ++j) {
+            if (residual.getComponent(j) < 0) {
+                int tmp = k[j];
+                signed[j] = (~tmp) & mask;
+            } else {
+                signed[j] = k[j];
+            }
+        }
+
+        return new QuantizeExResult(signed, q.t, q.ipNormInv);
+    }
+
+    /**
+     * Method to quantize a vector.
+     *
+     * @param oAbs   absolute values of a L2-normalized residual vector (nonnegative; length = dim)
+     * @return       quantized levels (ex-bits), the chosen scale t, and ipNormInv
+     * Notes:
+     * - If the residual is the all-zero vector (or numerically so), this returns zero codes,
+     *   t = 0, and ipNormInv = 1 (benign fallback).
+     * - Downstream code (ex_bits_code_with_factor) uses ipNormInv to compute f_rescale_ex, etc.
+     */
+    private QuantizeExResult quantizeEx(@Nonnull final RealVector oAbs) {
+        final int dim = oAbs.getNumDimensions();
+        final int maxLevel = (1 << numExBits) - 1;
+
+        // Choose t via the sweep.
+        double t = bestRescaleFactor(oAbs);
+        // ipNorm = sum_i ( (k_i + 0.5) * |r_i| )
+        double ipNorm = 0.0;
+
+        // Build per-coordinate integer levels: k_i = floor(t * |r_i|)
+        int[] code = new int[dim];
+        for (int i = 0; i < dim; i++) {
+            int k = (int) Math.floor(t * oAbs.getComponent(i) + EPS);
+            if (k > maxLevel) {
+                k = maxLevel;
+            }
+            code[i] = k;
+            ipNorm += (k + 0.5) * oAbs.getComponent(i);
+        }
+
+        // ipNormInv = 1 / ipNorm, with a benign fallback.
+        double ipNormInv;
+        if (ipNorm > 0.0 && Double.isFinite(ipNorm)) {
+            ipNormInv = 1.0 / ipNorm;
+            if (!Double.isFinite(ipNormInv) || ipNormInv == 0.0) {
+                ipNormInv = 1.0; // extremely defensive
+            }
+        } else {
+            ipNormInv = 1.0; // fallback used in the C++ source
+        }
+
+        return new QuantizeExResult(code, t, ipNormInv);
+    }
+
+    /**
+     *  Method to compute the best factor {@code t}.
+     *  @param oAbs   absolute values of a (row-wise) normalized residual; length = dim; nonnegative
+     *  @return t     the rescale factor that maximizes the objective
+     */
+    private double bestRescaleFactor(@Nonnull final RealVector oAbs) {
```

Comment:
This method is a bit confusing. I think I know what it's trying to accomplish (namely, that it's trying to get a rescale factor that places the vector nearest to one of the quantized code points), but I'm not sure I follow how it actually does that.